README
Django-Geoportail - Geodjango with the maps and photos from the French
National Geographic Institute.

 * Installation: see INSTALL
 * Get involved: http://bitbucket.org/bruno/django-geoportail/
 * Bugs: http://bitbucket.org/bruno/django-geoportail/issues/

Quickstart
==========

Add some maps to your django sites...


Known Issues
------------

The maps are not displayed properly for Multi* fields:

 * MultiPointField
 * MultiLineStringField
 * MultiPolygonField
 * (untested) GeometryCollectionField

In those cases, the objects will be displayed on the map but the map won't be
automatically focused to the objects when you try to edit existing objects.
You'll have to find them manually. Editing works.


In the admin
------------

To register a geographic model, add those lines to your <app_dir>/admin.py:

    from geoportal.admin import GeoPortalAdmin
    from django.contrib.gis import admin
    from my_app.models import MyModel

    admin.site.register(MyModel, GeoPortalAdmin)

If you need some customization, you can easily override the GeoPortalAdmin
class:

    class MyGeoAdmin(GeoPortalAdmin):
        # Your options here

Available options are:

 * ``map_width``: the width of the map (in pixels). Default: 600

 * ``map_height``: the height of the map, still in pixels. Default: 400

 * ``default_zoom``: the zoom level when a new empty map is created. Default: 5

 * ``default_lon``: default longitude for a new map. Default: 2

 * ``default_lat``: default latitude for a new map. Default: 47.
   ``default_lon``, ``default_lat`` and ``default_zoom`` are set to display a
   map centered on France with the whole country.

 * ``point_zoom``: the zoom level to select when you see a single point.
   Default: 15, for a 1:25000 map (best resolution available)

 * ``map_info`` (broken atm): show Geoportal's scale and coordinates widget.
   Default: True

 * ``layers``: customize the layers. Default: 'auto'. Could be either:

   * 'auto': automatically fill the map with a combination of the available
     layers (maps, photos...).
   * A 2-tuple tuple: ``( ('code', opacity), (..., ...), )``. ``code`` is the
     codename of the layer you want to display, ``opacity`` is its opacity,
     between 0 and 1.

     The order is important, the layers are added to the map in the same order
     as they are defined. If you add a layer with an opacity set to 1 to some
     previously added layers, this will hide the previous layers.

     Available layers are:

      * 'photos'
      * 'maps'
      * 'terrain': elevation map
      * 'cadaster': cadastral parcels
      * 'hydrography'
      * 'roads'
      * 'railways'
      * 'runways'
      * 'buildings'
      * 'gov': utility and governmental services
      * 'boundaries': adminstrative boundaries
      * 'coast': sea regions

     Note that layers may or may not be available depending on your API key.
     The free contract gives you only 'photos' and 'maps'.

     For more information about the different layers, please visit:
     https://api.ign.fr/geoportail/api/doc/fr/webmaster/layers.html


In your templates
-----------------

(not yet)


CHANGES
2010-11-22 Bruno Renié <bruno@renie.fr> -- 0.4.5

    * Fixed an issue with styleMap having effects on other maps when there are
      several widgets on the same page.

2010-08-24 Bruno Renié <bruno@renie.fr> -- 0.4.4

    * Update to Geoportail 1.1
    * Fixed a bug related to USE_L10N and language settings

2010-06-02 Bruno Renié <bruno@renie.fr>

    * #3 - Various things no longer break with French installations

2010-06-01 Bruno Renié <bruno@renie.fr> -- 0.4.3

    * #2 - Popups are enabled by default on KML and GPX insertion.
    * #1 - Setting the opacity to a float value no longer crashes French
      django installations.

2010-05-11 Bruno Renié <bruno@renie.fr> -- 0.4.2

    * KML and GPX inclusion got an autofocus parameter.

2010-05-10 Bruno Renié <bruno@renie.fr> -- 0.4.1

    * docs are included in the package data

2010-05-09 Bruno Renié <bruno@renie.fr> -- 0.4

    * Added KML and GPX rendering to the template library
    * Tests passes with Django 1.2

2010-05-08 Bruno Renié <bruno@renie.fr>

    * Zooming to the correct bounds when features are set to 'invisible'
    * Fixed the 'as map_var' behaviour in the templates
    * Added transitionEffect: 'resize' to all widgets for better UX
    * Fixed setup.py again
    * Removed hardcoded SRID in the admin widget

2010-05-03 Bruno Renié <bruno@renie.fr> -- 0.3.1

    * Fixed setup.py and released 0.3.1
    * Added a form library with rich widgets and relevant docs

2010-05-01 Bruno Renié <bruno@renie.fr>

    * Fixed cross-browser issues by adding a 'geoportal_js' tag

2010-04-24 Bruno Renié <bruno@renie.fr> -- 0.2.3

    * Released 0.2.3

2010-04-08 Bruno Renié <bruno@renie.fr>

    * Changed map_info to be false by default
    * Switched to the hosted version of the JS API
    * Added the PanZoomBar control to all maps in the admin widget
    * Updated the docs and screenshots accordingly

2010-03-03 Bruno Renié <bruno@renie.fr> -- 0.2.2

    * Added some tests for the template rendering
    * Javascript code for maps is completely isolated
    * Added a 'as var_name' to the template tag for further manipulation
    * Added tests (and a bare models.py to prevent the tests to fail)
    * Small fixes in the docs

2010-03-02 Bruno Renié <bruno@renie.fr> -- 0.2.1

    * Fix in the javascript: allowing more than 2 maps per page
    * More robust argument parsing, allowing template variables
    * Added version number at the root of the package
    * Docs -- online at http://packages.python.org/django-geoportail/


2010-03-02 Bruno Renié <bruno@renie.fr> -- 0.2.0

    * Added a template tag for rendering geo* fields
    * No more request to api.ign.fr, going straight to the token server


2010-01-08 Bruno Renié <bruno@renie.fr> -- 0.1.0

    * First release, admin support only



geoportal/templatetags/geoportal_tags.py
from django import template
from django.conf import settings
from django.utils.safestring import mark_safe

from django.contrib.gis.gdal import OGRGeomType

from geoportal import utils

import random

register = template.Library()


class SafeVariable(template.Variable):

    def resolve(self, context):
        try:
            return super(SafeVariable, self).resolve(context)
        except template.VariableDoesNotExist:
            return self.var


class OptionsNode(template.Node):
    """A template node with a nice option parsing bahaviour."""
    args = 0  # Number of required arguments
    available_options = ()  # List of options the parser should recognize

    def parse_options(self, args):
        self.options = {}
        if len(args) > self.args:
            # Eating empty options, u''
            options = [o for o in ''.join(args[self.args:]).split(',') if o]
            for o in options:
                key, value = o.split('=')
                if not key in self.available_options:
                    raise template.TemplateSyntaxError('"%s" option is not su'
                            'pported. Available options are: %s' % (key,
                            ', '.join(self.available_options)))
                self.options[key] = SafeVariable(value)

    def to_boolean(self, var_name):
        try:
            self.options[var_name] = bool(int(self.options[var_name]))
        except ValueError:
            raise template.TemplateSyntaxError('"%s" can be either 0 or 1 (was'
                ': "%s")' % (var_name, self.options[var_name]))


class MapNode(OptionsNode):
    args = 2
    available_options = ('width', 'height', 'visible', 'color',
                         'opacity', 'zoom', 'navigation')

    def __init__(self, args, var_name=None):
        self.geo_field = template.Variable(args[1])
        self.var_name = var_name
        self.args_ = args

    def render(self, context):
        self.parse_options(self.args_)
        # Generate a probably unique name for javascript variables -- in case
        # there are several maps on a page
        map_var = ''.join(random.sample('abcdefghijklmopqrstuvwxyz', 5))
        map_var = 'map_' + map_var

        # Field type
        geo_field = self.geo_field.resolve(context)
        ftype = geo_field.geom_type.upper()
        is_collection = ftype in ('MULTIPOINT', 'MULTILINESTRING',
                                  'MULTIPOLYGON', 'GEOMETRYCOLLECTION')
        if is_collection:
            if ftype == 'GEOMETRYCOLLECTION':
                collection_type = 'Any'
            else:
                collection_type = OGRGeomType(ftype.replace('MULTI', ''))
        else:
            collection_type = 'None'

        for (key, value) in self.options.items():
            self.options[key] = value.resolve(context)

        # Default options
        if not 'width' in self.options:
            self.options['width'] = getattr(settings,
                                            'GEOPORTAL_DEFAULT_WIDTH',
                                            utils.DEFAULT_WIDTH)

        if not 'height' in self.options:
            self.options['height'] = getattr(settings,
                                             'GEOPORTAL_DEFAULT_HEIGHT',
                                             utils.DEFAULT_HEIGHT)

        if not 'color' in self.options:
            self.options['color'] = getattr(settings,
                                            'GEOPORTAL_DEFAULT_COLOR',
                                            utils.DEFAULT_COLOR)

        if not 'opacity' in self.options:
            self.options['opacity'] = getattr(settings,
                                              'GEOPORTAL_DEFAULT_OPACITY',
                                              utils.DEFAULT_OPACITY)
        self.options['opacity'] = str(self.options['opacity'])

        self.check_booleans()

        isolated_context = template.Context({
            'options': self.options,
            'api_key': settings.GEOPORTAL_API_KEY,
            'map_var': map_var,
            'is_point': ftype in ('POINT', 'MULTIPOINT'),
            'is_linestring': ftype in ('LINESTRING', 'MULTILINESTRING'),
            'is_polygon': ftype in ('POLYGON', 'MULTIPOLYGON'),
            'is_collection': is_collection,
            'collection_type': collection_type,
            'layers': utils.get_layers((('maps', 1),)),
            'default_lon': getattr(settings, 'GEOPORTAL_DEFAULT_LON',
                                   utils.DEFAULT_LON),
            'default_lat': getattr(settings, 'GEOPORTAL_DEFAULT_LAT',
                                   utils.DEFAULT_LAT),
            'default_zoom': getattr(settings, 'GEOPORTAL_DEFAULT_ZOOM',
                                    utils.DEFAULT_ZOOM),
            'point_zoom': getattr(settings, 'GEOPORTAL_POINT_ZOOM',
                                  utils.POINT_ZOOM),
            'srid': 4326,
            'field_name': ftype.capitalize(),
            'wkt': geo_field.wkt,
            'wms_url': utils.WMS_URL,
        })
        loaded = template.loader.get_template('geoportal/map.html')
        rendered = loaded.render(isolated_context)
        if self.var_name is not None:
            context[self.var_name] = map_var
        return rendered

    def check_booleans(self):
        if not 'visible' in self.options:
            self.options['visible'] = 1
        else:
            self.to_boolean('visible')

        if 'navigation' in self.options:
            self.to_boolean('navigation')


@register.tag
def geoportal_map(parser, token):
    """
    {% geoportal_map field option1=value1, option2=value2, ... %}

    Renders a geographic field (point, polygon, path...)

    Options:             Default
     * width (pixels)    400
     * height (pixels)   300
     * visible (1|0)     1
     * color (rrggbb)    ee9900 (OpenLayers.Feature.Vector
                                 .style["default"]["fillColor"])
     * opacity (0 -> 1)  0.4
     * zoom (~3 -> 14)   if not provided, calculated automatically
     * navigation (1|0)  turn navigation on
    """
    bits = token.split_contents()
    if len(bits) < 2:
        raise template.TemplateSyntaxError('geoportal_map takes at least one '
                                           'argument')
    if len(bits) > 3 and bits[-2] == 'as':
        return MapNode(bits[:-2], var_name=bits[-1])
    return MapNode(bits)


@register.simple_tag
def geoportal_js():
    media_url = getattr(settings, 'GEOPORTAL_MEDIA_URL', utils.MEDIA_URL)
    return mark_safe('<script type="text/javascript" src=' + \
                     '"%sGeoportalExtended.js"></script>' % media_url)


class GmlNode(OptionsNode):
    args = 3
    available_options = ('color', 'opacity', 'width', 'extract',
                         'focus', 'popup')
    format_type = None
    force_style = False

    def __init__(self, args):
        self.map_var = template.Variable(args[1])
        self.kml_url = SafeVariable(args[2])
        self.parse_options(args)

    def render(self, context):
        for key, value in self.options.items():
            self.options[key] = value.resolve(context)

        if self.force_style:
            self.options['extract'] = True
            self.options['style'] = True
        else:
            if 'extract' in self.options:
                self.to_boolean('extract')
            else:
                self.options['extract'] = True

            for key in ('width', 'opacity', 'color'):
                # Setting any of those disables style extraction
                if key in self.options:
                    self.options['extract'] = False
                    break

        if not 'width' in self.options:
            self.options['width'] = 2

        if not 'opacity' in self.options:
            if self.format_type == 'gpx':
                self.options['opacity'] = 1
            else:
                self.options['opacity'] = utils.DEFAULT_OPACITY
        self.options['opacity'] = str(self.options['opacity'])

        if not 'color' in self.options:
            self.options['color'] = utils.DEFAULT_COLOR

        if 'focus' in self.options:
            self.to_boolean('focus')
        else:
            self.options['focus'] = True

        if 'popup' in self.options:
            self.to_boolean('popup')
        else:
            self.options['popup'] = True

        self.options['url'] = self.kml_url.resolve(context)
        self.options['map_var'] = self.map_var.resolve(context)
        self.options['format'] = self.format_type
        return template.loader.render_to_string('geoportal/gml.html',
                                                self.options)


class KmlNode(GmlNode):
    format_type = 'kml'


@register.tag
def geoportal_kml(parser, token):
    """Adds a layer to an existing map from a KML file
    {% geoportal_kml map_var kml_url color=..., width=..., opacity=... %}

    Options all have default values:
        color: utils.DEFAULT_COLOR
        width: 2
        opacity: utils.DEFAULT_OPACITY
        extract: 1
    """
    bits = token.split_contents()
    if len(bits) < 3:
        raise template.TemplateSyntaxError('geoportal_kml takes at least two '
                                           'arguments')
    return KmlNode(bits)


class GpxNode(GmlNode):
    format_type = 'gpx'
    force_style = True
    available_options = ('color', 'opacity', 'width', 'focus', 'popup')


@register.tag
def geoportal_gpx(parser, token):
    """Same as KML, renders a GPX file.
    Options (defaut value):
        width: width of the feature (2)
        opacity: stroke opacity (1)
        color: feature color (utils.DEFAULT_COLOR)
    """
    bits = token.split_contents()
    if len(bits) < 3:
        raise template.TemplateSyntaxError('geoportal_gpx takes at least two '
                                           'arguments')
    return GpxNode(bits)
